Firebase/Messaging/FIRMessagingRemoteNotificationsProxy.m

```diff
@@ -526,25 +545,106 @@ void FCM_swizzle_willPresentNotificationWithHandler(
     return;
   }
 
-  // Valid original method signature, go ahead to swizzle.
+  // Attempt to access the user info
+  id notificationUserInfo = userInfoFromNotification(notification);
+
+  if (!notificationUserInfo) {
+    // Could not access notification.request.content.userInfo.
+    callOriginalMethodIfAvailable();
+    return;
+  }
+
+  [[FIRMessaging messaging] appDidReceiveMessage:notificationUserInfo];
+  // Execute the original implementation.
+  callOriginalMethodIfAvailable();
+}
+
+/**
+ * Swizzle the notification handler for iOS 10+ devices.
+ * Signature of original handler is as below:
+ * - (void)userNotificationCenter:(UNUserNotificationCenter *)center
+ *     didReceiveNotificationResponse:(UNNotificationResponse *)response
+ *     withCompletionHandler:(void (^)(void))completionHandler
+ * In order to make FCM SDK compile and compatible with iOS SDKs before iOS 10, hide the
+ * parameter types from the swizzling implementation.
+ */
+void FCM_swizzle_didReceiveNotificationResponseWithHandler(
+    id self, SEL _cmd, id center, id response, void (^handler)()) {
+
+  FIRMessagingRemoteNotificationsProxy *proxy = [FIRMessagingRemoteNotificationsProxy sharedProxy];
+  IMP original_imp = [proxy originalImplementationForSelector:_cmd];
+
+  void (^callOriginalMethodIfAvailable)() = ^{
+    if (original_imp) {
+      ((void (*)(id, SEL, id, id, void (^)(void)))original_imp)(
+          self, _cmd, center, response, handler);
+    }
+    return;
+  };
+
+  Class notificationCenterClass = NSClassFromString(@"UNUserNotificationCenter");
+  Class responseClass = NSClassFromString(@"UNNotificationResponse");
+  if (!center || ![center isKindOfClass:[notificationCenterClass class]]) {
+    // Invalid parameter type from the original method.
+    // Do not swizzle, just execute the original method.
+    callOriginalMethodIfAvailable();
```

Comment:
out of scope of this change, but why do we call the original method when attempting to swizzle? shouldn't swizzling just be writing class state?

Reply:
The idea here is that we're just making an additional call to handle messaging analytics, and _always_ calling the original method, so the developer everything behaves as if nothing had been swizzled.

Because of the runtime-reflection-based calling of everything in UNUserNotificationCenter (because we still have to be compatible with building on Xcode 7), there are a _lot_ more checks of things that could (but will almost never) go wrong.

In those extreme cases, we just call the original method and do nothing extra. It's likely that if something was wrong here, it'll probably just blow up when it gets back to the happy path too, but we should just be as pass-through as possible, if that makes sense.